pypesto/objective/roadrunner/roadrunner_calculator.py

```diff
@@ -0,0 +1,403 @@
+                    mapping_species[key] = float(value)
+            else:
+                mapping_params[key] = value
+
+        if filling_mode == "only_parameters" or filling_mode == "all":
+            # set parameters.
+            roadrunner_instance.setValues(mapping_params)
+            # reset is necessary to apply the changes to initial assignments
+            roadrunner_instance.reset()
+        if filling_mode == "only_species" or filling_mode == "all":
+            # set species
+            roadrunner_instance.setValues(mapping_species)
+        return mapping_values
+
+
+def calculate_llh(
+    simulations: np.ndarray,
+    edata: ExpData,
+    parameter_mapping: dict,
+) -> float:
+    """Calculate the negative log-likelihood for a single condition.
+
+    Parameters
+    ----------
+    simulations:
+        Simulations of condition.
+    edata:
+        ExpData of a single condition.
+    parameter_mapping:
+        Parameter mapping for the condition.
+
+    Returns
+    -------
+    float:
+        Negative log-likelihood.
+    """
+    # if 0 is not in timepoints, remove the first row of the simulation
+    if 0.0 not in edata.timepoints:
+        simulations = simulations[1:, :]
+
+    if not np.array_equal(simulations[:, 0], edata.timepoints):
+        raise ValueError(
+            "Simulation and Measurement have different timepoints."
+        )
+    # check that simulation and condition have same dimensions and timepoints
+    if simulations.shape != edata.measurements.shape:
+        raise ValueError(
+            "Simulation and Measurement have different dimensions."
+        )
+    # we can now drop the timepoints
+    simulations = simulations[:, 1:]
+    measurements = edata.measurements[:, 1:]
+
+    def _fill_in_noise_formula(noise_formula):
+        """Fill in the noise formula."""
+        if isinstance(noise_formula, numbers.Number):
+            return float(noise_formula)
+        # if it is not a number, it is assumed to be a string
+        if noise_formula in parameter_mapping.keys():
+            return parameter_mapping[noise_formula]
+
+    # replace noise formula with actual value from mapping
+    noise_formulae = np.array(
+        [_fill_in_noise_formula(formula) for formula in edata.noise_formulae]
+    )
+    # check that the rows of noise are the columns of the simulation
+    if noise_formulae.shape[0] != simulations.shape[1]:
+        raise ValueError("Noise and Simulation have different dimensions.")
+    # per observable, decide on the llh function based on the noise dist
+    llhs = np.array(
+        [
+            LLH_TYPES[noise_dist](
+                measurements[:, i], simulations[:, i], noise_formulae[i]
+            )
+            for i, noise_dist in enumerate(edata.noise_distributions)
+        ]
+    ).transpose()
+    # check whether all nan values in llhs coincide with nan measurements
+    if not np.all(np.isnan(llhs) == np.isnan(measurements)):
+        return np.nan
```

Comment:
Should there be a warning here? First trying to see if there are more or less nan values in the llh. And then in the first case warning that there are additional nan values in the llh calculation, or the opposite? (Not sure if the second is possible, but first can probably occur)

Reply:
I am not sure what you mean 🤔  if we encounter a nan-value in llh, there are two options: in the first case, it is also an nan value in the measurements - that's fine, we don't really care, as it indicates that there was no measurement to begin with, in the other case it is not, and independent of who has more nan values, this should result in a nan value of the objective, as the simulation corresponding to a valid measurement was not possible. That was at least my thought process here. Happy for suggestions :)

Reply:
True, my mistake. I misunderstood a part of it, but still have a question.
I was concerned with the case when the llh is nan where a measurement is not nan. This can happen if the simulation is nan for that specific non-nan measurement. In that case, it's the job of the simulator to warn that there are nan values.

But I'm asking if there is a case when both simulation and measurement are non-nan, but the llh is nan. In this case, would you want to pass a warning here that you saw a nan value?

Definitely not important in the big picture and ok as it is.